metaconfig-typesafe-config/src/main/scala/metaconfig/typesafeconfig/TypesafeConfig2Class.scala

```diff
@@ -53,36 +53,73 @@ object TypesafeConfig2Class {
               )
           }
       }
-      getPositionOpt(value.origin, input) match {
+      getPositionOpt(value.origin, cache) match {
         case Some(pos) => conf.withPos(pos)
         case None => conf
       }
     }
     try {
-      Configured.Ok(loop(config().resolve().root))
+      val resolved = config().resolve().root()
+      cache += OriginId(resolved.origin) -> input
+      Configured.Ok(loop(resolved))
     } catch {
       case e: ConfigException.Parse =>
         Configured.NotOk(
-          ConfError.parseError(getPosition(e.origin, input), e.getMessage)
+          ConfError.parseError(getPosition(e.origin, cache), e.getMessage)
         )
     }
   }
 
   private def getPosition(
       originOrNull: ConfigOrigin,
-      input: Input
+      cache: mutable.Map[OriginId, Input]
   ): Position =
-    getPositionOpt(originOrNull, input).getOrElse(Position.None)
+    getPositionOpt(originOrNull, cache).getOrElse(Position.None)
 
   private def getPositionOpt(
       originOrNull: ConfigOrigin,
-      input: Input
+      cache: mutable.Map[OriginId, Input]
   ): Option[Position] =
     for {
       origin <- Option(originOrNull)
-      linePlus1 <- Option(origin.lineNumber)
-      line = linePlus1 - 1
-      start = input.lineToOffset(line)
+      input <- cache.updateWith(OriginId(origin)) { maybeCurrent =>
+        maybeCurrent
+          .orElse(
+            Option(origin.url)
+              .map(url => Input.File(new java.io.File(url.toURI)))
+          )
+      }
+      linePlus1 = origin.lineNumber
+      start = input.lineToOffset(linePlus1 - 1)
     } yield Position.Range(input, start, start)
 
+  private case class OriginId private (id: String)
+  private object OriginId {
+    def apply(origin: ConfigOrigin): OriginId =
+      OriginId(
+        origin
+          .withLineNumber(-1) // strip potential line suffix
+          // use description instead of the nullable filename() or url(),
+          // so that we can have an identifier to map to the main input
+          // when using ConfigFactory.parseString(string)
```

Comment:
the complexity of the new implementation comes from `gimmeConfFromStringFilename`, present in `TypesafeConfig2Class` but not in `Sconfig2Class`, and tested via https://github.com/scalameta/metaconfig/blob/cf69d53586e6389f66b06db864df7e776a9a96bb/metaconfig-tests/jvm/src/test/scala/metaconfig/cli/EchoOptionsSuite.scala#L7-L30